clang/lib/CIR/Dialect/Transforms/MergeCleanups.cpp

```diff
@@ -80,6 +80,22 @@ struct RemoveEmptySwitch : public OpRewritePattern<SwitchOp> {
   }
 };
 
+struct RemoveRedundantBitcast : OpRewritePattern<CastOp> {
+  using OpRewritePattern<CastOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(CastOp op,
+                                PatternRewriter &rewriter) const final {
+    if (op.getKind() != mlir::cir::CastKind::bitcast) {
+      return failure();
+    }
+    if (op.getSrc().getType() != op.getResult().getType()) {
+      return failure();
+    }
+    rewriter.replaceOp(op, op.getSrc());
+    return success();
```

Comment:
This looks great, thanks! As you noticed, we don't use the canonicalizer just yet (there are issues tracking it, see https://github.com/llvm/clangir/issues/542 and https://github.com/llvm/clangir/issues/543), in a gist the reason is that it's too aggressive and we need to fix CIR operations to better incorporate side effects.

I wonder if we can write these in terms of fold's though, so instead we could improve and reuse `CastOp::fold` here such that when we fix the canonicalizer bugs, we can just delete `RemoveRedundantBitcast` and fallback to the canonicalizer. Have you tried anything in that direction?

Reply:
Oh, I see. This option sounds much more reasonable. As `fold` is already performed by `applyOpPatternsAndFold`, I can also remove the pattern now.